ansible_base/rbac/api/serializers.py

```diff
@@ -0,0 +1,198 @@
+
+    def to_representation(self, value):
+        raise NotImplementedError
+
+    def __init__(self, **kwargs):
+        # Workaround so that the parent class does not resolve the choices right away
+        self.html_cutoff = kwargs.pop('html_cutoff', self.html_cutoff)
+        self.html_cutoff_text = kwargs.pop('html_cutoff_text', self.html_cutoff_text)
+
+        self.allow_blank = kwargs.pop('allow_blank', False)
+        super(serializers.ChoiceField, self).__init__(**kwargs)
+
+    def _initialize_choices(self):
+        choices = self.get_dynamic_choices()
+        self._grouped_choices = to_choices_dict(choices)
+        self._choices = flatten_choices_dict(self._grouped_choices)
+        self.choice_strings_to_values = {str(k): k for k in self._choices}
+
+    @property
+    def grouped_choices(self):
+        if not hasattr(self, '_grouped_choices'):
+            self._initialize_choices()
+        return self._grouped_choices
+
+    @property
+    def choices(self):
+        if not hasattr(self, '_choices'):
+            self._initialize_choices()
+        return self._choices
+
+    def to_internal_value(self, data):
+        try:
+            return self.get_dynamic_object(data)
+        except ObjectDoesNotExist:
+            self.fail('does_not_exist', pk_value=data)
+        except (TypeError, ValueError):
+            self.fail('incorrect_type', data_type=type(data).__name__)
+
+
+class ContentTypeField(ChoiceLikeMixin):
+    psuedo_field = 'model'
+
+    def __init__(self, **kwargs):
+        kwargs['help_text'] = _('The type of resource this applies to')
+        super().__init__(**kwargs)
+
+    def get_model_for_init(self):
+        return permission_registry.content_type_model
+
+    def get_dynamic_choices(self):
+        return [
+            (f'{settings.ANSIBLE_BASE_SERVICE_PREFIX}.{cls._meta.model_name}', cls._meta.verbose_name.title())
+            for cls in permission_registry.all_registered_models
+        ]
+
+    def get_dynamic_object(self, data):
+        model = data.rsplit('.')[-1]
+        return permission_registry.content_type_model.objects.get(model=model)
+
+    def to_representation(self, value):
+        return f'{settings.ANSIBLE_BASE_SERVICE_PREFIX}.{value.model}'
+
+
+class PermissionField(ChoiceLikeMixin):
+    psuedo_field = 'codename'
+
+    def get_model_for_init(self):
+        return permission_registry.permission_model
+
+    def get_dynamic_choices(self):
+        perms = []
+        for cls in permission_registry.all_registered_models:
+            cls_name = cls._meta.model_name
+            for action in cls._meta.default_permissions:
+                perms.append(f'{settings.ANSIBLE_BASE_SERVICE_PREFIX}.{action}_{cls_name}')
+            for perm_name, description in cls._meta.permissions:
+                perms.append(f'{settings.ANSIBLE_BASE_SERVICE_PREFIX}.{perm_name}')
+        return perms
+
+    def get_dynamic_object(self, data):
+        codename = data.rsplit('.')[-1]
```

Comment:
Don't remove the app label from the model name. It can result in this error if multiple apps have a model with teh same name: 

```
pulp-1          |   File "/src/django-ansible-base/ansible_base/rbac/api/serializers.py", line 67, in to_internal_value
pulp-1          |     return self.get_dynamic_object(data)
pulp-1          |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
pulp-1          |   File "/src/django-ansible-base/ansible_base/rbac/api/serializers.py", line 92, in get_dynamic_object
pulp-1          |     return permission_registry.content_type_model.objects.get(model=model)
pulp-1          |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
pulp-1          |   File "/usr/local/lib/python3.11/site-packages/django/db/models/manager.py", line 87, in manager_method
pulp-1          |     return getattr(self.get_queryset(), name)(*args, **kwargs)
pulp-1          |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
pulp-1          |   File "/usr/local/lib/python3.11/site-packages/django/db/models/query.py", line 640, in get
pulp-1          |     raise self.model.MultipleObjectsReturned(
```

Reply:
The `data` isn't giving a Django permission reference in the format `"{app_name}.{codename}"` like you are thinking. The DAB resource registry app introduced model references along the lines of `"{service_name}.{model_name}"` like `"awx.inventory"`. This is using permission references according to `"{service_name}.{codename}"`, like `"awx.change_inventory"`, and yes, this requires unique codenames among the usage of this, just like the resource registry does for model names. That's not saying you can't, for example, have proxy models. But you cannot track multiple proxied models in the resource registry, and you can't do it with this RBAC app either. Right now, this only operate for a single service.

I am absolutely not changing this, this is driven by much broader commitments that I have no control over. You can't get an app_name because there is no app_name to obtain. This is a model interchange format that goes between multiple different services. AWX, for instance uses `auth.User` and eda-server uses `core.User`. The interchange format treats these as the same model.

If this is really a problem, see what AWX and eda-server already have with the `DABPermission` model, which uses utils from here to only populate the entries with models tracked in the permission_registry. That works very well. If that can't work with pulp, then the query here can be narrowed to the registered content types.

Reply:
@AlanCoding Ah, my bad. I misread the code. Do you have a fix for this? I'm guess it' was caused by all the group proxy models we have.

Reply:
Yes, I pushed something I believe to be a fix, neglected to update this comment chain. Model `ProxyInventory` was introduced (not actually a proxy model) which abstractly is supposed to cover what you reported here.

This feedback is incredible btw. Test models and tests are introduced as these come in.